rfd/9999-passwordless-fido2.md

```diff
@@ -0,0 +1,551 @@
+during build, or once again forking go-libfido2 in order to standardize building
+and linking strategies. (Static linking seems to be the best option for `tsh`.)
+
+During development, the libfido2-aware code in Teleport is to be protected by a
+`libfido2` build tag - said tag could be kept in the final product if dealing
+with the library proves to be a hassle.
+
+### Security
+
+Security considerations are largely unchanged in relation to the
+[Passwordless RFD][passwordless rfd].
+
+It's important to note that CLI applications with hardware key access are
+particularly dangerous, as they are not subject to the security measures
+we've come to expect from browsers (in fact, _any_ untrusted application that
+interacts with hardware keys should be used with caution).
+Those concerns are not exclusive to WebAuthn, but are made more delicate in face
+of passwordless credentials.
+
+`tsh` does its best to be well-behaved and mitigates hardware key access
+concerns by being open source, providing signed binaries and its own Cloud
+solution.
+
+(Security considerations for the [browser-based CLI authentication](
+#browser-based-cli-authentication) alternative section are contained within it.)
+
+### UX
+
+UX is discussed throughout the design.
+
+## Alternatives considered
+
+### Browser-based CLI authentication
+
+The diagram below exemplifies a new login method: the assignee based login.
+In this method a third-party, the assignee, performs the authentication ceremony
+on behalf of different entity (ie, Web UI authenticates on behalf of tsh).
+
+The design is largely based in a new set of RPCs and removes the need for
+communication between the source and assignee, apart from the initial stimulus.
+
+```
+                                           tsh authentication via browser
+
+       ┌─┐
+       ║"│
+       └┬┘
+       ┌┼┐
+        │                 ┌───┐                        ┌────────────────┐                             ┌────────┐
+       ┌┴┐                │tsh│                        │Browser / Web UI│                             │Teleport│
+      user                └─┬─┘                        └───────┬────────┘                             └───┬────┘
+       │ tsh login --pwdless│                                  │                                          │
+       │ ───────────────────>                                  │                                          │
+       │                    │                                  │                                          │
+       │                    │                   CreateAssigneeCredentials(sshPublicKey)                   │
+       │                    │ ────────────────────────────────────────────────────────────────────────────>
+       │                    │                                  │                                          │
+       │                    │                                  assigneeID                                 │
```

Comment:
Wouldn't this facilitate MITM attacks? It seems that the `assigneeID` secret is a single point of failure for the ceremony. What is preventing an attacker from serving their own `/web/assignee?assigneeID=n` links and phish a victim?

Reply:
I think there are a few protecting factors here:

1) `tsh` is the sole holder of the private key, fishing for the assignee ID won't give you that (the key never leaves tsh)
2) To MITM `/web/assignee` you need to impersonate the Teleport Proxy the user talks to. Assuming a proper prod setup (ie, no `tsh login --insecure`) this is hopefully hard to do.

 It's also important to note that the assignee ID is not really a secret, although it is hard to guess beforehand.

Reply:
I'm sorry, I could have been clearer. In the theorized attack scenario:
* An adversary `tsh` starts authentication by registering a new assignee login in Teleport, including the public key to be signed.
* The victim is tricked into completing the authentication ceremony using the resulting assignee ID in their web browser, who performs the authentication ceremony on the evil `tsh`'s behalf.
* Once the ceremony is complete, the evil `tsh` retrieves the credentials from Teleport.

Is the above scenario reasonable or am I reading the design flow incorrectly/missing something?

Reply:
An adversary `tsh` that manages to trick the user could run the login ceremony by itself as well, regardless of passwordless or assignee logins, but I think I get your point now.

What I think we could do to mitigate is to make it extra clear, in the Web UI, that we are running login on behalf of `tsh` and show all the relevant information (binary name, user, origin IP, so on, as captured by the Teleport servers).

Do you any ideas to change or secure the flow? This is an optional bit of the design at the moment, but I do foresee it being useful for Terminal.

Reply:
I feel that we'll need to open a dedicated RFC just for this alternative, since a new threat modeling process is needed to identify the right mitigations for this flow. One of the security needs I can anticipate is the need for implementing something similar to the OAuth Proof Key for Code Exchange pattern. PKCE ([RFC 7636](https://datatracker.ietf.org/doc/html/rfc7636)) is an extension to the Authorization Code flow that would prevent certain attacks and that is needed to securely perform the OAuth exchange from public clients. The technique involves the client first creating a secret, hashing it, and then using that secret again when exchanging the authorization code for an access token. This way if the code is intercepted, it will not be useful since the token request relies on the initial secret. See https://auth0.com/docs/get-started/authentication-and-authorization-flow/authorization-code-flow-with-proof-key-for-code-exchange-pkce#how-it-works for a complete diagram of the flow. Depending on the implementation, `tsh` may need to implement a similar proof to get the credentials.

Reply:
Thanks for the detailed reply, Lorenzo. I'll catch up with the links and do any follow up questions later.

No objections on my part for taking this section out. We can move forward without it and cross the Terminal bridge when we have to (maybe they can move to "pure" Web authentication or get away with using `tsh` flows).

Reply:
Removed section from the design. We can revisit in a future design, if at all.

Reply:
@codingllama for `Teleport Terminal`, as it's an electron app, can we simply use the `web` flow using `credentials provider` without any pop ups?